### PR TITLE
S3 -> Spark and SparkSQL

### PR DESCRIPTION
**Pull request by cpcloud**
Adds support for loading S3 csv and JSON into Spark and SparkSQL

### CODE REVIEW

File: into/backends/sparksql.py [hunk 1]
```diff
@@ -57,7 +65,16 @@ def rdd_to_sqlcontext(ctx, rdd, name=None, dshape=None, **kwargs):
...
     sql_schema = dshape_to_schema(dshape)
-    sdf = ctx.applySchema(rdd, sql_schema)
+
+    try:
+        sdf = ctx.applySchema(rdd, sql_schema)
+    except TypeError:
+        # TODO: gross
+        # we have to convert a tuple of strings to have typed elements
+        # this seems like it would be horribly inefficient
+        rdd = rdd.map(lambda x: convert(np.ndarray, [x], dshape=dshape).item())
```
Comment (cpcloud) [inline into/backends/sparksql.py]: i wonder if we could use `dynd` here

File: into/backends/csv.py [hunk 2]
```diff
@@ -277,4 +278,13 @@ def infer_header(csv, encoding='utf-8', **kwargs):
             not all(dt == 'O' for dt in df.dtypes))
 
 
+def get_delimiter(csv, length=8192):
+    """Figure out the delimiter of a CSV file.
+    """
+    sniffer = Sniffer()
+    with sample(csv, length=length) as fn:
+        with open(fn, mode='rb') as f:
+            return sniffer.sniff(f.read()).delimiter
```
Comment (mrocklin) [inline into/backends/csv.py]: Suggest generalizing to `get_dialect` which returns a dict with all of the delimiter terms

File: into/backends/spark.py [hunk 3]
```diff
@@ -32,6 +34,11 @@ def anything_to_spark_context(sc, o, **kwargs):
     return append(sc, convert(list, o, **kwargs), **kwargs)
 
 
+@append.register(RDD, RDD)
+def append_rdd_to_rdd(a, b, **kwargs):
+    return a.union(b)
```
Comment (mrocklin) [inline into/backends/spark.py]: We should be aware that this breaks conventions of `append` (which are pretty loose anyway) as an operation that mutates the left hand side.

File: into/backends/spark.py [hunk 4]
```diff
@@ -49,3 +56,32 @@ def discover_rdd(rdd, n=50, **kwargs):
...
+
+
+@append.register(SparkContext, S3(CSV))
+def s3_csv_to_rdd(sc, s3, dshape=None, use_unicode=False, minPartitions=None,
+                  **kwargs):
+    """Convert an S3(CSV) file to a Spark RDD.
+
+    Notes
+    -----
+    Performs no type inference.
```
Comment (mrocklin) [inline into/backends/spark.py]: What would happen if I did

    into(my_sql_context, S3(CSV)(...))

File: into/backends/sparksql.py (hunk 1 above)
Comment (mrocklin) [inline into/backends/sparksql.py]: We used to.  Also consider `rdd.mapPartitions` which will do this same operation but in bulk across a list of tuples at a time.

Comment (mrocklin) [inline into/backends/sparksql.py]: Note also that this would fail if the workers don't have into installed which may very well be the case.

Comment (mrocklin) [inline into/backends/sparksql.py]: Is SparkSQL able to make these coercions?

In the future when there is more time we could also consider make a Pure Python function that transformed a datashape into a function to do a conversion like this.  I'd want to wait until someone asked for this before spending time on it though.

File: into/backends/sparksql.py [hunk 5]
```diff
@@ -46,6 +48,12 @@ def jsonlines_to_sparksql(ctx, json, dshape=None, name=None, schema=None,
     return srdd
 
 
+@append.register(SQLContext, S3(JSONLines))
```
Comment (mrocklin) [inline into/backends/sparksql.py]: This is cool.

**Comment by cpcloud** [comment]
I'd like to wait to merge this after the next spark release. Their new `DataFrame` API fixes some of my database inspection hacks and has some breaking changes with out code that I'd like to address.

**Comment by mrocklin** [comment]
What's the status here?  We're waiting on Spark 1.3?  Presumably this doesn't go in to the next release then?

**Comment by cpcloud** [comment]
yep i don't think this will go in the next release

**Comment by cpcloud** [comment]
closing. will revisit later.

**Pull request closed without merging**